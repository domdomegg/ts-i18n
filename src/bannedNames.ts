export const bannedNames = [
  "AggregateError",
  "Array",
  "ArrayBuffer",
  "AsyncFunction",
  "AsyncGenerator",
  "AsyncGeneratorFunction",
  "Atomics",
  "BigInt",
  "BigInt64Array",
  "BigUint64Array",
  "Boolean",
  "DataView",
  "Date",
  "Error",
  "EvalError",
  "Float32Array",
  "Float64Array",
  "Function",
  "Generator",
  "GeneratorFunction",
  "Infinity",
  "Int16Array",
  "Int32Array",
  "Int8Array",
  "InternalError",
  "Intl",
  "JSON",
  "Map",
  "Math",
  "NaN",
  "Number",
  "Object",
  "Promise",
  "Proxy",
  "RangeError",
  "ReferenceError",
  "Reflect",
  "RegExp",
  "Set",
  "SharedArrayBuffer",
  "String",
  "Symbol",
  "SyntaxError",
  "TypeError",
  "URIError",
  "Uint16Array",
  "Uint32Array",
  "Uint8Array",
  "Uint8ClampedArray",
  "WeakMap",
  "WeakSet",
  "WebAssembly",
  "abstract",
  "arguments",
  "await",
  "boolean",
  "break",
  "byte",
  "case",
  "catch",
  "char",
  "class",
  "const",
  "continue",
  "debugger",
  "decodeURI",
  "decodeURIComponent",
  "default",
  "delete",
  "do",
  "double",
  "else",
  "encodeURI",
  "encodeURIComponent",
  "enum",
  "escape",
  "eval",
  "export",
  "extends",
  "false",
  "final",
  "finally",
  "float",
  "for",
  "function",
  "globalThis",
  "goto",
  "if",
  "implements",
  "import",
  "in",
  "instanceof",
  "int",
  "interface",
  "isFinite",
  "isNaN",
  "let",
  "long",
  "native",
  "new",
  "null",
  "package",
  "parseFloat",
  "parseInt",
  "private",
  "protected",
  "public",
  "return",
  "short",
  "static",
  "super",
  "switch",
  "synchronized",
  "this",
  "throw",
  "throws",
  "transient",
  "true",
  "try",
  "typeof",
  "undefined",
  "unescape",
  "uneval",
  "var",
  "void",
  "volatile",
  "while",
  "with",
  "yield",
];
