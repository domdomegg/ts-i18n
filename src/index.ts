export { generateCore, GenerateCoreOptions } from './core';
export { generateFs, GenerateFsOptions } from './fsGenerator';
