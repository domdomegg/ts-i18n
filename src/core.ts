import bannedNames from './bannedNames';

type LanguageType = { [key: string]: LanguageTypeKey }
type LanguageTypeKey = { type: 'object', values: LanguageType } | { type: 'string', params: Set<string> }

type LanguageDefinition = { [key: string]: LanguageDefinitionKey }
type LanguageDefinitionKey = LanguageDefinition | string

export type Files = { name: string, content: string }[]

export const generateCore = (options: {
  inputFiles: Files,
  defaultLanguage: string,
}): Files => {
  const outputFiles: Files = [];

  if (new Set(options.inputFiles.map((f) => f.name)).size !== options.inputFiles.length) {
    throw new Error('Duplicate input file name');
  }

  const defaultLanguageFile = options.inputFiles.find((f) => f.name.slice(0, -5) === options.defaultLanguage);
  if (!defaultLanguageFile) {
    throw new Error(`Default language file does not exist (expected ${options.defaultLanguage}.json)`);
  }

  const languageType: LanguageType = options.inputFiles.reduce<LanguageType>((acc, file) => {
    if (!file.name.endsWith('.json')) throw new Error(`Input file names must end with .json, but received ${file.name}`);
    const parsedContent = asLanguageDefinition(saferJSONParse(file.content, file.name), file.name);

    const { code, languageType } = generateCodeAndLanguageType(parsedContent, file === defaultLanguageFile, acc);
    outputFiles.push({ name: `${file.name.slice(0, -5)}.ts`, content: code });
    return languageType;
  }, {});

  // TODO: check the default language meets the language type (we might get this for free with TS? not sure - need to check)
  // i.e. assertThat(generateLanguageType(defaultLanguageFile)).isSubTypeOf(languageType)

  outputFiles.push({ name: 'types.ts', content: generateTypes(languageType, false) });
  outputFiles.push({ name: 'utils.ts', content: generateUtils(options) });
  outputFiles.push({ name: 'browser.ts', content: generateBrowser() });
  outputFiles.push({ name: '.gitignore', content: `${outputFiles.map((f) => f.name).join('\n')}\n.gitignore` });

  return outputFiles;
};

const saferJSONParse = (str: string, errorIn?: string) => {
  try {
    return JSON.parse(str);
  } catch {
    throw new Error(`Failed to pass JSON${errorIn ? ` (in ${errorIn})` : ''}`);
  }
};

// Checks an object is a valid language definition, throwing if it isn't one
const asLanguageDefinition = (obj: unknown, errorIn?: string): LanguageDefinition => {
  if (Array.isArray(obj)) {
    throw new Error(`Invalid language definition as found array${errorIn ? ` (in ${errorIn})` : ''}`);
  }

  if (typeof obj !== 'object') {
    throw new Error(`Invalid language definition as found ${typeof obj}${errorIn ? ` (in ${errorIn})` : ''}`);
  }

  Object.entries(obj as { [key: string]: unknown }).forEach(([key, value]) => {
    if (!/[a-zA-Z_$][0-9a-zA-Z_$]*/.test(key)) throw new Error(`Invalid language definition due to bad key: '${key}'${errorIn ? ` (in ${errorIn})` : ''}`);

    if (typeof value !== 'string') {
      asLanguageDefinition(value, errorIn);
    }
  });

  return obj as LanguageDefinition;
};

/**
 * @param language The parsed language definition
 * @param isDefault Whether this is the default language
 * @param languageType The constraints on the language type. May be modified by this function.
 * @returns code for the given language definition, and updated constraints on the language type
 */
// TODO: possibly extract the language type generation
const generateCodeAndLanguageType = (language: LanguageDefinition, isDefault: boolean, languageType: LanguageType = {}): { code: string, languageType: LanguageType } => ({
  code: `// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import { Language${isDefault ? '' : ', DeepPartial'} } from "./types"
const lang = {${indent(generateCodeFragment(language, languageType))}\n}
export default lang as ${(isDefault ? 'Language' : 'DeepPartial<Language>')}\n`,
  languageType,
});

/**
 * @param language A language definition
 * @param languageType The constraints on the language type. May be modified by this function.
 * @param path Path in the language definition we are at if this is a recursive call, to improve error messaging
 * @returns fragment of code for the given language definition, and updated constraints on the language type
 */
// TODO: general tidy up, and possibly extract the language type generation
/* eslint-disable no-param-reassign */
const generateCodeFragment = (language: LanguageDefinition, languageType: LanguageType, path = '') => {
  let code = '';
  Object.entries(language).forEach(([key, languageValue]) => {
    if (typeof languageValue === 'object') {
      if (languageType[key] === undefined) {
        languageType[key] = { type: 'object', values: {} };
      }

      const languageTypeKey = languageType[key];
      if (languageTypeKey.type !== 'object') {
        throw new Error(`Incompatible type at ${path}, some translation files define objects and others define strings`);
      }

      code += `\n${key}: {${indent(generateCodeFragment(languageValue, languageTypeKey.values, path.length ? `${path}[${JSON.stringify(key)}]` : JSON.stringify(key)))}\n},`;
    } else {
      if (languageType[key] === undefined) {
        languageType[key] = { type: 'string', params: new Set() };
      }

      const languageTypeKey = languageType[key];
      if (languageTypeKey.type !== 'string') {
        throw new Error(`Incompatible type at ${path}, some translation files define objects and others define strings`);
      }

      const stringified = JSON.stringify(languageValue);

      // TODO: come up with a way of escaping parameters
      const params = [...new Set([...stringified.matchAll(/\{\{[a-zA-Z_$][0-9a-zA-Z_$]*\}\}/g)].map((match) => match[0].slice(2, -2)))];
      params.forEach((param) => {
        if (bannedNames.includes(param)) throw new Error(`Cannot use param name '${param}' as it is a reserved word`);
      });
      if (params.length && `${key}_plural` in language) {
        const countRequired = params.includes('count');
        if (!countRequired) params.push('count');

        const body = `p.count === 1 ? (${stringified.replace(/\{\{([a-zA-Z_$][0-9a-zA-Z_$]*)\}\}/g, '" + p.$1.toString() + "')}) : lang.${path}${key}_plural(p)`;
        code += `\n${key}: (p: { ${params.map((p) => (p === 'count' ? `${p + (countRequired ? '' : '?')}: number` : `${p}: string | number`)).join(', ')} }): string => ${body},`;
      } else if (params.length) {
        const body = stringified.replace(/\{\{([a-zA-Z_$][0-9a-zA-Z_$]*)\}\}/g, '" + p.$1.toString() + "');
        code += `\n${key}: (p: { ${params.map((p) => `${p}: string | number`).join(', ')} }): string => ${body},`;
      } else {
        code += `\n${key}: (): string => ${stringified},`;
      }

      params.forEach((param) => languageTypeKey.params.add(param));
    }
  });
  return code;
};

const generateTypes = (languageType: LanguageType, includePlurals = true) => (
  `${'// Do not edit directly, this is generated by ts-i18n\n'
  + '/* eslint-disable */\n'
  + 'export type DeepPartial<T> = { [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P]; }\n'
  + 'export interface Language {'}${indent(generateTypeFragment(languageType, includePlurals))}\n}\n`
);

const generateTypeFragment = (languageType: LanguageType, includePlurals: boolean) => {
  let code = '';
  Object.entries(languageType).forEach(([key, languageTypeKey]) => {
    if (languageTypeKey.type === 'object') {
      code += `\n${key}: {${indent(generateTypeFragment(languageTypeKey.values, includePlurals))}\n},`;
    } else if (!includePlurals && key.endsWith('_plural')) {
      // noop
    } else if (languageTypeKey.params.size === 0) {
      code += `\n${key}: () => string,`;
    } else {
      code += `\n${key}: (p: { ${[...languageTypeKey.params].map((p) => (p === 'count' ? `${p}: number` : `${p}: string | number`)).join(', ')} }) => string,`;
    }
  });
  return code;
};

const generateUtils = (options: { inputFiles: Files, defaultLanguage: string }) => `// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
${options.inputFiles.map((f) => `import ${f.name.slice(0, -5)} from './${f.name.slice(0, -5)}'\n`).join('')}import { Language, DeepPartial } from './types'

/**
 * Merges two languages together, preferring the preferred language where possible
 * @param preferredLang The preferred language. NB: may be modified
 * @param defaultLang The default language. NB: may be modified
 */
export const merge = (preferredLang: DeepPartial<Language>, defaultLang: Language): Language => {
  for (const key in defaultLang) {
    if (key in preferredLang) {
      if (typeof (preferredLang as any)[key] === 'object') {
        (preferredLang as any)[key] = merge((preferredLang as any)[key], (defaultLang as any)[key])
      }
    } else {
      (preferredLang as any)[key] = (defaultLang as any)[key]
    }
  }
  return preferredLang as Language
}

export const languages = { ${options.inputFiles.map((f) => f.name.slice(0, -5)).join(', ')} }

export const defaultLanguage = ${options.defaultLanguage}

export const defaultLanguageCode = ${JSON.stringify(options.defaultLanguage)}`;

const generateBrowser = () => `// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import { merge, languages, defaultLanguage } from './generated/index'
import { Language } from './generated/types'

// Preferred languages, as specified by the user's browser
const preferredLanguages =
  ((navigator.languages ? navigator.languages : [navigator.language]) ?? [])
    .flatMap(lang => lang.length > 2 ? [lang, lang.slice(0, 2)] : lang)
    .filter((l, i, arr) => arr.indexOf(l) === i)
    .slice(0, 25)

let language: Language = defaultLanguage // fallback language
for (let i = preferredLanguages.length; i > 0; i--) {
  const key = preferredLanguages[i] as keyof typeof languages
  if (key in languages) {
    language = merge(languages[key], language)
  }
}

export default language
`;

const indent = (str: string): string => str.split('\n').map((line) => (line.length ? `  ${line}` : line)).join('\n');
