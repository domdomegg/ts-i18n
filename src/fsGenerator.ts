import { mkdirSync, readdirSync, readFileSync, writeFileSync } from "fs";
import { resolve } from "path";
import { generateCore, Files } from "./core";

export const generateFs = (options: {
  inputDirectory: string;
  outputDirectory: string;
  defaultLanguage: string;
}): void => {
  createDirectory(options.outputDirectory);
  const languageFiles = getLanguageFiles(options.inputDirectory);
  if (!languageFiles.length)
    throw new Error("No language files found in " + options.inputDirectory);
  const code = generateCore({
    inputFiles: languageFiles,
    defaultLanguage: options.defaultLanguage,
  });
  writeToDirectory(options.outputDirectory, code);
};

const createDirectory = (directory: string): void => {
  mkdirSync(resolve(directory), { recursive: true });
};

const getLanguageFiles = (directory: string): Files => {
  return readdirSync(resolve(directory))
    .filter((f) => f.endsWith(".json"))
    .map((f) => ({
      name: f,
      content: readFileSync(resolve(directory, f), { encoding: "utf-8" }),
    }));
};

const writeToDirectory = (directory: string, files: Files): void => {
  for (const file of files) {
    writeFileSync(resolve(directory, file.name), file.content, {
      encoding: "utf-8",
    });
  }
};
