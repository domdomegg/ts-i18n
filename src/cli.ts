#!/usr/bin/env node

import { generateFs } from "./fsGenerator";
import { program } from "commander";

const cli = (args: string[]) => {
  program
    .description("i18n with proper TypeScript support")
    .requiredOption(
      "-i, --input <folder>",
      "Input folder for your input JSON files"
    )
    .requiredOption(
      "-o, --output <folder>",
      "Output folder for your generated .ts files"
    )
    .requiredOption(
      "-d, --default-language <lang>",
      "The default/fallback/primary language",
      "en"
    )
    .parse(args);

  generateFs({
    inputDirectory: program.opts().input,
    outputDirectory: program.opts().output,
    defaultLanguage: program.opts().defaultLanguage,
  });
};

try {
  cli(process.argv);
} catch (e) {
  // A bold red 'Error: '
  const errorPrefix = "\x1b[1;31mError:\x1b[0m ";

  const helpHint = "\x1b[1mHint:\x1b[0m View help with --help";

  if (e instanceof Error) {
    console.error(errorPrefix + e.message);
    console.error(helpHint);
    process.exit(1);
  } else {
    console.error(errorPrefix + e);
    console.error(helpHint);
    process.exit(2);
  }
}
