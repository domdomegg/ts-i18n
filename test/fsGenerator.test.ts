import { readdirSync, readFileSync, rmdirSync, existsSync } from 'fs';
import { resolve } from 'path';
import { generateFs } from '../src/fsGenerator';

test.each(['simple', 'large'])('fsGenerator generates correct files: %s', (dir) => {
  if (existsSync(resolve('test/generated'))) {
    rmdirSync(resolve('test/generated'), { recursive: true });
  }

  generateFs({
    inputDirectory: `test/resources/${dir}`,
    outputDirectory: 'test/generated',
    defaultLanguage: 'en',
  });

  expect(readdirSync(resolve(__dirname, './generated/'))).toHaveLength(4);
  expect(readFileSync(resolve(__dirname, './generated/en.ts'), { encoding: 'utf-8' })).toMatchSnapshot();
  expect(readFileSync(resolve(__dirname, './generated/fr.ts'), { encoding: 'utf-8' })).toMatchSnapshot();
  expect(readFileSync(resolve(__dirname, './generated/types.ts'), { encoding: 'utf-8' })).toMatchSnapshot();
  expect(readFileSync(resolve(__dirname, './generated/index.ts'), { encoding: 'utf-8' })).toMatchSnapshot();
});
