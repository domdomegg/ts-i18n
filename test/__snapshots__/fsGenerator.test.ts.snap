// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fsGenerator generates correct files: large 1`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import { Language } from \\"./types\\"
const lang = {
  yes: (): string => \\"Yes\\",
  no: (): string => \\"No\\",
  thereAreXFeatures: (p: { count: number }): string => p.count === 1 ? (\\"There is \\" + p.count.toString() + \\" feature\\") : lang.thereAreXFeatures_plural(p),
  thereAreXFeatures_plural: (p: { count: string | number }): string => \\"There are \\" + p.count.toString() + \\" features\\",
  nested: {
    works: {
      correctly: (): string => \\"Nested works correctly\\",
      really: {
        well: (): string => \\"Nested works really well\\",
      },
    },
  },
  andYouCanMissKeys: (): string => \\"And you can miss keys in non-default languages\\",
  orHaveDifferentPlaceholders: (p: { firstName: string | number }): string => \\"Or have different placeholders, \\" + p.firstName.toString() + \\"\\",
  manyParameters: (p: { title: string | number, firstName: string | number, lastName: string | number }): string => \\"You can have many parameters, \\" + p.title.toString() + \\" \\" + p.firstName.toString() + \\" \\" + p.lastName.toString() + \\"!\\",
}
export default lang as Language
"
`;

exports[`fsGenerator generates correct files: large 2`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import { Language, DeepPartial } from \\"./types\\"
const lang = {
  yes: (): string => \\"Oui\\",
  no: (): string => \\"Non\\",
  thereAreXFeatures: (p: { count: number }): string => p.count === 1 ? (\\"Il y a \\" + p.count.toString() + \\" fonctionnalité\\") : lang.thereAreXFeatures_plural(p),
  thereAreXFeatures_plural: (p: { count: string | number }): string => \\"Il y a \\" + p.count.toString() + \\" fonctionnalités\\",
  nested: {
    works: {
      correctly: (): string => \\"Imbriqué fonctionne correctement\\",
      really: {
        well: (): string => \\"Imbriqué fonctionne très bien\\",
      },
    },
  },
  orHaveDifferentPlaceholders: (p: { username: string | number }): string => \\"Ou avoir des espaces réservés différents, \\" + p.username.toString() + \\"\\",
}
export default lang as DeepPartial<Language>
"
`;

exports[`fsGenerator generates correct files: large 3`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
export type DeepPartial<T> = { [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P]; }
export interface Language {
  yes: () => string,
  no: () => string,
  thereAreXFeatures: (p: { count: number }) => string,
  nested: {
    works: {
      correctly: () => string,
      really: {
        well: () => string,
      },
    },
  },
  andYouCanMissKeys: () => string,
  orHaveDifferentPlaceholders: (p: { firstName: string | number, username: string | number }) => string,
  manyParameters: (p: { title: string | number, firstName: string | number, lastName: string | number }) => string,
}
"
`;

exports[`fsGenerator generates correct files: large 4`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import en from './en'
import fr from './fr'
import { Language, DeepPartial } from './types'

/**
 * Merges two languages together, preferring the preferred language where possible
 * @param preferredLang The preferred language. NB: may be modified
 * @param defaultLang The default language. NB: may be modified
 */
export const merge = (preferredLang: DeepPartial<Language>, defaultLang: Language): Language => {
  for (const key in defaultLang) {
    if (key in preferredLang) {
      if (typeof (preferredLang as any)[key] === 'object') {
        (preferredLang as any)[key] = merge((preferredLang as any)[key], (defaultLang as any)[key])
      }
    } else {
      (preferredLang as any)[key] = (defaultLang as any)[key]
    }
  }
  return preferredLang as Language
}

export const languages = { en, fr }

export const defaultLanguage = \\"en\\""
`;

exports[`fsGenerator generates correct files: simple 1`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import { Language } from \\"./types\\"
const lang = {
  thereAreXFeatures: (p: { count: number }): string => p.count === 1 ? (\\"There is \\" + p.count.toString() + \\" feature\\") : lang.thereAreXFeatures_plural(p),
  thereAreXFeatures_plural: (p: { count: string | number }): string => \\"There are \\" + p.count.toString() + \\" features\\",
  nested: {
    works: (): string => \\"Nested works\\",
  },
  andYouCanMissKeys: (): string => \\"And you can miss keys in non-default languages\\",
  orHaveDifferentPlaceholders: (p: { firstName: string | number }): string => \\"Or have different placeholders, \\" + p.firstName.toString() + \\"\\",
}
export default lang as Language
"
`;

exports[`fsGenerator generates correct files: simple 2`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import { Language, DeepPartial } from \\"./types\\"
const lang = {
  thereAreXFeatures: (p: { count: number }): string => p.count === 1 ? (\\"Il y a \\" + p.count.toString() + \\" fonctionnalité\\") : lang.thereAreXFeatures_plural(p),
  thereAreXFeatures_plural: (p: { count: string | number }): string => \\"Il y a \\" + p.count.toString() + \\" fonctionnalités\\",
  nested: {
    works: (): string => \\"Imbriqué fonctionne\\",
  },
  orHaveDifferentPlaceholders: (p: { username: string | number }): string => \\"Ou avoir des espaces réservés différents, \\" + p.username.toString() + \\"\\",
}
export default lang as DeepPartial<Language>
"
`;

exports[`fsGenerator generates correct files: simple 3`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
export type DeepPartial<T> = { [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P]; }
export interface Language {
  thereAreXFeatures: (p: { count: number }) => string,
  nested: {
    works: () => string,
  },
  andYouCanMissKeys: () => string,
  orHaveDifferentPlaceholders: (p: { firstName: string | number, username: string | number }) => string,
}
"
`;

exports[`fsGenerator generates correct files: simple 4`] = `
"// Do not edit directly, this is generated by ts-i18n
/* eslint-disable */
import en from './en'
import fr from './fr'
import { Language, DeepPartial } from './types'

/**
 * Merges two languages together, preferring the preferred language where possible
 * @param preferredLang The preferred language. NB: may be modified
 * @param defaultLang The default language. NB: may be modified
 */
export const merge = (preferredLang: DeepPartial<Language>, defaultLang: Language): Language => {
  for (const key in defaultLang) {
    if (key in preferredLang) {
      if (typeof (preferredLang as any)[key] === 'object') {
        (preferredLang as any)[key] = merge((preferredLang as any)[key], (defaultLang as any)[key])
      }
    } else {
      (preferredLang as any)[key] = (defaultLang as any)[key]
    }
  }
  return preferredLang as Language
}

export const languages = { en, fr }

export const defaultLanguage = \\"en\\""
`;
